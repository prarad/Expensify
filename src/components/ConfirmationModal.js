import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { closeModal } from '../actions/confirmationModal';
import { startRemoveExpense } from '../actions/expenses';

export const ConfirmationModal = ({
  modalIsOpen,
  handleRequestClose,
  startRemoveExpense,
  selectedExpense = {},
  history
}) => (
    <Modal
      className="modal"
      isOpen={modalIsOpen}
      contentLabel="Confirmation Modal"
      onRequestClose={handleRequestClose}
      closeTime={200}
      ariaHideApp={false}
    >
      <h2 className="modal__title">Confirmation</h2>
      <p className="modal__body">
        Are you sure to delete '{<span>{selectedExpense.description}</span>}' item?
    </p>
      <div className="modal__answers">
        <button
          className="btn"
          onClick={() => {
            startRemoveExpense(selectedExpense.id)
              .then(() => {
                handleRequestClose();
                history.push('/');
              })
          }}
        >Yes</button>
        <button className="btn" onClick={handleRequestClose}>No</button>
      </div>
    </Modal>
  );

const mapStateToProps = state => ({
  modalIsOpen: state.ModalIsOpen
});

const mapDispatchToProps = dispatch => ({
  handleRequestClose: () => dispatch(closeModal()),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(ConfirmationModal);