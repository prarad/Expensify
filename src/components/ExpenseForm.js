import React from 'react';
import { addExpense } from '../actions/expenses'
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
    createdAt: moment(),
    calendarFocused: false,
    error: ''
  }
  dispatch = this.props.dispatch;

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  }
  onDateChange = createdAt => {
    createdAt &&
      this.setState(() => ({ createdAt }))
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount'
      }))
    } else {
      this.setState(() => ({ error: '' }))
    }
    !this.state.error && this.dispatch(
      addExpense({
        description: this.state.description,
        note: this.state.note,
        amount: this.state.amount,
        createdAt: this.state.createdAt
      })
    )
  }
  render() {
    return (
      <div className='Expense-Form'>

        {!!this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input type="text"
            placeholder='Description'
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder='Amount'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea name="" id="" cols="30" rows="10"
            placeholder='Add a note for your expense(optional)'
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div >
    )
  }
}