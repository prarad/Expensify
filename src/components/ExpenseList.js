import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = props => (
  <div className="container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Expenses</span>
        </div>
      ) : (
          props.expenses.map(expense => (
            <ExpenseListItem {...expense} key={expense.id} />
          )))
      }
    </div>
  </div>
)

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)