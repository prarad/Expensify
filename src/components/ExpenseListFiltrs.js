import React from 'react';
import uuid from 'uuid'
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }))
  }
  onTextChange = e => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = e => {
    e.target.value === 'date' ?
      this.props.sortByDate() :
      this.props.sortByAmount()
  }
  render() {
    return (
      <div className="container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search expenses"
            />
          </div>
          <div className="input-group__item">
            <select
              name="sort-by"
              className="select"
              value={this.props.filters.text}
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            {this.props.filters.sortBy === 'date' &&
              (
                <DateRangePicker
                  id='date-range-picker'
                  startDate={this.props.filters.startDate}
                  startDateId='start-date-id'
                  endDate={this.props.filters.endDate}
                  endDateId='end-date-id'
                  onDatesChange={this.onDatesChange}
                  focusedInput={this.state.calenderFocused}
                  onFocusChange={this.onFocusChange}
                  showClearDates={true}
                  numberOfMonths={1}
                  isOutsideRange={() => false}
                />
              )
            }
          </div>
        </div>
      </div>
    )
  }
}
const mapStateToProps = state => ({ filters: state.filters })
const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)