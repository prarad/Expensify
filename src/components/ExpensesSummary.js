import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

const expenseWordGenerator = base => base === 1 ? 'expense' : 'exepenses';

export const ExpensesSummary = ({ expensesCount, expensesTotal, hiddenExpenses }) => {
  const expenseWord = expenseWordGenerator(expensesCount);
  const hiddenExpenseWord = expenseWordGenerator(hiddenExpenses);
  const expensesTotalFormatted = numeral(expensesTotal / 100).format('$0,0.00')
  return (
    <div className='page-header'>
      <div className="container">
        <h1 className="page-header__title">Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{expensesTotalFormatted}</span></h1>
        {!!hiddenExpenses && <h4 className="page-header__subtitle">{`${hiddenExpenses} hidden ${hiddenExpenseWord}`}</h4>}
        <div className="page-header__actions">
          <Link className="btn" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
    hiddenExpenses: state.expenses.length - visibleExpenses.length,
  }
}

export default connect(mapStateToProps)(ExpensesSummary)