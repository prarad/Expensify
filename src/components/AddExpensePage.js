import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

const AddExpensePage = ({ dispatch, history }) => (
  <div className='AddExpensePage'>
    <h1>Add Expense</h1>
    <ExpenseForm
      action='add'
      onSubmit={(expense => {
        dispatch(addExpense(expense))
        history.push('/')
      })} />
  </div>
)

export default connect()(AddExpensePage)