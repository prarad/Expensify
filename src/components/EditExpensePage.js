import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense({
      id: this.props.expense.id,
      updates: expense
    });
    this.props.history.push('/')
  }
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id)
    this.props.history.push('/')
  }
  render() {
    return (
      < div >
        <div className="page-header">
          <div className="container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button
            className="btn btn--secondary"
            onClick={this.onRemove}
          >Remove Expense</button>
        </div>
      </div >
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startRemoveExpense: id => dispatch(startRemoveExpense(id)),
  startEditExpense: ({ id, updates }) => dispatch(startEditExpense({ id, updates }))
})

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)