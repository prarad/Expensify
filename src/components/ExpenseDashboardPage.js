import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFiltrs';

export default () => (
  <div className="ExpenseDashboardPage">
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)