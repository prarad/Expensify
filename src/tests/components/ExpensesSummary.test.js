import React from 'react';
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should correctly render ExpensesSumarry with 1 expenes', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={39} />)
  expect(wrapper).toMatchSnapshot()
})

test('should correctly render ExpensesSumarry with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={3} expensesTotal={31396} />)
  expect(wrapper).toMatchSnapshot()
})