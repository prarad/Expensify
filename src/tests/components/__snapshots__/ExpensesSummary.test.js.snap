// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSumarry with 1 expenes 1`] = `
<div
  className="page-header"
>
  <div
    className="container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $0.39
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="btn"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSumarry with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $313.96
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="btn"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
